### PR TITLE
Add general pubsub message

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### What was wrong?

Address #63 by adding a generic PubSub message type.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

**Comment by mhchia** [comment]
Note: we can also restructure request messages from Go to Python(e.g. `GetCollation`). E.g. on Go side, the request message only needs to be `peerID`, `bytes`, and probably an indicator `type` to indicate what the format of `bytes` is. Then, request messages will be deserialized, handled in Python, and send the result(`response`, `is_message_valid`) back to Go.

**Comment by mhchia** [comment]
Note:
TODO
- [ ] change code to send data through new RPCs
- [ ] remove redundant code

**Comment by mhchia** [comment]
Note: this PR works with the Python code https://github.com/mhchia/sharding-python

File: main_test.go [hunk 1]
```diff
@@ -458,37 +449,50 @@ func makeEventResponse(success bool) *pbevent.Response {
...
+	default:
+		success = false
 	}
-	res := &pbevent.GetCollationResponse{
-		Response:  makeEventResponse(true),
-		Collation: collation,
-	}
-	return res, nil
+	return &pbevent.ReceiveResponse{
+		Response: makeEventResponse(success),
```
Comment (mhchia) [inline main_test.go]: Should check whether we can return "invalidity" through `error`, to avoid HTTP-like response code

File: main_test.go [hunk 2]
```diff
@@ -512,6 +516,27 @@ func runEventServer(ctx context.Context, eventRPCAddr string) (*eventTestServer,
...
+func TestEventRPCPubSub(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	nodes := makeNodes(t, ctx, 2)
+	connectBarbell(t, ctx, nodes)
+	shardID := ShardIDType(1)
+	nodes[0].ListenShard(ctx, shardID)
+	nodes[1].ListenShard(ctx, shardID)
+	waitForPubSubMeshBuilt()
+	eventRPCPort := 35566
```
Comment (mhchia) [inline main_test.go]: Haven't run `runEventServer`, this `eventRPC` should fail by all means

File: main_test.go [hunk 3]
```diff
@@ -531,27 +556,38 @@ func TestEventRPCNotifier(t *testing.T) {
...
+	reqBytes := protoToBytes(t, req)
+	respBytes, err := eventNotifier.Receive("", typeCollationRequest, reqBytes)
+	if err != nil {
+		t.Errorf("something wrong when calling `eventNotifier.Receive`: %v", err)
+	}
+	collationReceived := &pbmsg.Collation{}
+	err = proto.Unmarshal(respBytes, collationReceived)
 	if err != nil {
-		t.Errorf("something wrong when calling `eventNotifier.GetCollation`: %v", err)
+		t.Errorf("failed to unmarshal collation bytes")
```
Comment (mhchia) [inline main_test.go]: seems it worth extracting these few lines to a function `requestCollation` for test, to avoid DRY

File: main_test.go [hunk 4]
```diff
@@ -646,10 +701,22 @@ func TestRequestCollationWithRPCEventNotifier(t *testing.T) {
...
+	reqBytes = protoToBytes(t, req)
+	respBytes, err := nodes[0].generalRequest(ctx, nodes[1].ID(), typeCollationRequest, reqBytes)
 	if err != nil {
 		t.Errorf("request collation failed: %v", err)
 	}
+	collation := &pbmsg.Collation{}
+	err = proto.Unmarshal(respBytes, collation)
+	if err != nil {
+		t.Errorf("failed to unmarshal collation bytes")
+	}
```
Comment (mhchia) [inline main_test.go]: ditto

File: main_test.go [hunk 5]
```diff
@@ -661,12 +728,17 @@ func TestRequestCollationWithRPCEventNotifier(t *testing.T) {
...
+		Hash:    "",
 	}
-	if collation != nil {
-		t.Errorf("collation should be nil because it it not found in the mock event server")
+	reqBytes = protoToBytes(t, req)
+	respBytes, err = nodes[0].generalRequest(ctx, nodes[1].ID(), typeCollationRequest, reqBytes)
+	if err == nil {
+		t.Errorf(
+			"should be unable to unmarshal because nodes[1] should send invalid collation, due to the lack of an eventNotifier",
+		)
```
Comment (mhchia) [inline main_test.go]: ditto

File: shardmanager.go [hunk 6]
```diff
@@ -31,6 +31,12 @@ type ShardManager struct {
 const listeningShardTopic = "listeningShard"
 const collationTopicFmt = "shardCollations_%d"
 
+const (
```
Comment (mhchia) [inline shardmanager.go]: This should be a temporary solution. We should not have knowledge of data type here

File: shardmanager.go [hunk 7]
```diff
@@ -303,37 +309,52 @@ func extractProtoMsg(pubsubMsg *pubsub.Message, msg proto.Message) error {
 	return nil
 }
 
-func (n *ShardManager) makeCollationValidator() TopicValidator {
+func (n *ShardManager) SubscribeCollation(ctx context.Context, shardID ShardIDType) error {
```
Comment (mhchia) [inline shardmanager.go]: Maybe it can be abstract to the upper layer as well

File: shardmanager.go [hunk 8]
```diff
@@ -381,17 +402,27 @@ func (n *ShardManager) broadcastCollation(
 	return nil
 }
 
+// FIXME: in this go layer, we shouldn't have knowledge that what we are broadcasting
+//		  This should be handled in the upper side, e.g. the Python host.
+// 		  However, this code makes us easier to test without spinning up the "remote host"
+//		  We should remove these "collation" specific messages when the host side is ready.
 func (n *ShardManager) broadcastCollationMessage(collation *pbmsg.Collation) error {
```
Comment (mhchia) [inline shardmanager.go]: ditto

File: main_test.go (hunk 2 above)
Comment (mhchia) [reply main_test.go]: Removed this test

File: main_test.go (hunk 3 above)
Comment (mhchia) [reply main_test.go]: IMO it seems not necessary for now, and can be refactored further later.

File: shardmanager.go (hunk 7 above)
Comment (mhchia) [reply shardmanager.go]: Refer to https://github.com/ethresearch/sharding-p2p-poc/issues/68

File: shardmanager.go (hunk 6 above)
Comment (mhchia) [reply shardmanager.go]: Added `FIXME`

**Comment by mhchia** [comment]
If we pass the CI, it seems we should try to ship this to the codebase, since https://github.com/ethereum/py-evm/pull/1311 depends on this.

**Comment by mhchia** [comment]
@ChihChengLiang 
If we pass the CI, I think we should try to ship this to the codebase, since https://github.com/ethereum/py-evm/pull/1311 depends on this. What do you think?

**Comment by ChihChengLiang** [comment]
That makes sense, is this ready for review?

File: main_test.go [hunk 9]
```diff
@@ -583,37 +573,49 @@ func makeEventResponse(success bool) *pbevent.Response {
...
+			resBytes, err = proto.Marshal(collation)
+			if err != nil {
+				success = false
+			}
+		} else {
+			success = false
+			// resBytes = []byte{} // return empty bytes to indicate that we don't have the collation
+		}
+	default:
+		success = false
```
Comment (ChihChengLiang) [inline main_test.go]: init `success` true and default here false is weird

Comment (ChihChengLiang) [reply main_test.go]: Oh ... I see how it works

Comment (mhchia) [reply main_test.go]: > init success true and default here false is weird

You're right I think. Will it be better if we just declare it without initializing to a specific value? like`var success bool`. I.e. it implies `success` is initialized to `false`

**Pull request merged**